src/WebJobs.Script/Description/CSharp/CSharpFunctionInvoker.cs

```diff
@@ -112,10 +114,10 @@ private void ReloadScript()
                 TraceWriter.Trace(traceEvent);
             }
 
-            // If the compilation succeeded, restart the host
-            // TODO: Make this smarter so we only restart the host if the 
-            // method signature (or types used in the signature) change.
-            if (!compilationResult.Any(d => d.Severity == DiagnosticSeverity.Error))
+            // If the compilation succeeded AND our function signature has changed, restart the host
+            if (!compilationResult.Any(d => d.Severity == DiagnosticSeverity.Error) &&
+                !_functionSignature.HasLocalTypeReference &&
```

Comment:
What does `HasLocalTypeReference` mean? Maybe capture that in the comment? It seems to disable the optimization.